### ISSUE TITLE
OrderedMultiDict with empty list.

### ISSUE THREAD

**Issue by dourvaris** [pallets/werkzeug#764]
OrderedMultiDict has issue with empty list item:

``` python
>>> MultiDict({'a': []}).to_dict(flat=False)
{'a': []} # correct output
>>> OrderedMultiDict({'a': []}).to_dict(flat=False)
{} # wrong output
```

**Comment by nateprewitt** [comment]
The issue that you're encountering seems to be a product of the [iter_multi_items](https://github.com/mitsuhiko/werkzeug/blob/master/werkzeug/datastructures.py#L40) function which is only yielding on iteration of the value. Since the empty list iterates 0 times, nothing is yielded which leaves the instantiated OrderedMultiDict empty. Below I've listed some of the hitches I've encountered try to work out a fix.  

We could add a "and len(value)" or "and len(value) > 0" check to line 40 which will trigger the else statement on line 43 and return 'a' and [] for key and value, respectively. Doing so, however, creates downstream issues in the add method, producing a final value of {'a': [[]]}. This is caused by the setdefault() call on [line 789](https://github.com/mitsuhiko/werkzeug/blob/master/werkzeug/datastructures.py#L789). 

I'm not aware of anything we can yield from iter_multi_items to use the add() call as it currently stands. If we add checks for an empty list in add(), it will be difficult to disambiguate between a genuine empty list element addition and a call produced by input like {'a': []}. From everything I've tried, there doesn't appear to be a way to add an element to the OrderedMultiDict without triggering the add() method. 

As it stands, the (semi)solutions I see are modifying add() to not add nested empty lists which will prevent intentional use of lists like [1, 'q', [], 4], or we document this issue with the class and write it off as an improbable occurrence with normal form data.

**Comment by dourvaris** [comment]
Currently multidict assumes that if you pass
```MultiDict({'a': [1, 2, 3]})``` you want `[('a', 1), ('a', 2), ('a', 3)]` - this differs from doing ```MultiDict()['a'] = [1, 2, 3]``` which results in ```[('a', [1, 2, 3])]```

I think the first functionality is useful, but that means that MultiDict({'a': []}) should in fact do nothing, or  raise an exception that lists cannot be empty.

Which means MultiDict gives the wrong output rather than OrderedMultiDict.

However if there was a special case made for empty lists/tuples to be added, I don't see an issue with `[[]]` being returned since you want a list of lists using flat=False. With `a` as an empty tuple for example you would expect `[()]`, not `[]` or `()`.

**Comment by davidism** [comment]
Seems to have been fixed in #991
